### PR TITLE
add pretty json output in marshal based on spec + options

### PR DESCRIPTION
**Pull request by Skytrias**
adds `Marshal_Options` to customize the marshal output

can switch between tabs / spaces with `use_spaces` & `spaces` >= 1
mjson options for not using quotes & using `=` instead of `:`
option to output `uint` as hex numbers

example using .MJSON
```odin
json.marshal(
	value, 
	{
		spec = .MJSON,
		pretty = true,
		write_uint_as_hex = true,
		mjson_keys_use_equal_sign = true,
	},
)
```

example using empty options
```odin
json.marshal(value, {})
```

### CODE REVIEW

File: core/encoding/json/marshal.odin [hunk 1]
```diff
@@ -333,3 +398,121 @@ marshal_to_writer :: proc(w: io.Writer, v: any) -> (err: Marshal_Error) {
...
+
+// writes current indentation level based on options
+opt_write_indentation :: proc(w: io.Writer, opt: ^Marshal_Options) -> (err: io.Error) {
+	if !opt.pretty {
+		return
+	}
+
+	// TODO optimize?
+	if opt.use_spaces {
+		// NOTE maybe max(1, opt.spaces)
```
Comment (gingerBill) [inline core/encoding/json/marshal.odin:507]: If `opt.spaces == 0`, default to either `1` or `4`

File: core/encoding/json/marshal.odin [hunk 2]
```diff
@@ -82,7 +111,23 @@ marshal_to_writer :: proc(w: io.Writer, v: any) -> (err: Marshal_Error) {
...
+		// allow uints to be printed as hex
+		if opt.write_uint_as_hex && (opt.spec == .JSON5 || opt.spec == .MJSON) {
+			switch i in a {
+				case u8, u16, u32, u64, u128: {
+					s = strconv.append_bits_128(buf[:], u, 16, info.signed, 8*ti.size, "0123456789abcdef", { .Prefix })
+				}
+
+				case: {
+					s = strconv.append_bits_128(buf[:], u, 10, info.signed, 8*ti.size, "0123456789", nil)
+				}
```
Comment (gingerBill) [inline core/encoding/json/marshal.odin:125]: Remove the indentation and `{}` for each `case` block.

File: core/encoding/json/marshal.odin [hunk 3]
```diff
@@ -333,3 +398,121 @@ marshal_to_writer :: proc(w: io.Writer, v: any) -> (err: Marshal_Error) {
...
+			} else {
+				io.write_string(w, name) or_return
+			}
+			
+			if opt.mjson_keys_use_equal_sign {
+				io.write_string(w, " = ") or_return
+			} else {
+				io.write_string(w, ": ") or_return
+			}
+		}
```
Comment (gingerBill) [inline core/encoding/json/marshal.odin:422]: Remove the indentation and {} for each case block.

File: core/encoding/json/marshal.odin [hunk 4]
```diff
@@ -333,3 +398,121 @@ marshal_to_writer :: proc(w: io.Writer, v: any) -> (err: Marshal_Error) {
...
+					io.write_byte(w, '\n') or_return
+				} else {
+					// NOTE comma seperation necessary for non pretty output!
+					io.write_string(w, ", ") or_return
+				}
+			}
+
+			opt_write_indentation(w, opt) or_return
+		}
+	}
```
Comment (gingerBill) [inline core/encoding/json/marshal.odin:474]: Remove the indentation and {} for each case block.

File: core/encoding/json/marshal.odin [hunk 5]
```diff
@@ -226,31 +267,55 @@ marshal_to_writer :: proc(w: io.Writer, v: any) -> (err: Marshal_Error) {
...
+							}
+
+							opt_write_key(w, opt, name) or_return
+						}
+	
+						case: {
+							// TODO better error output?
+							return .Unsupported_Type
+						}
+					}
```
Comment (gingerBill) [inline core/encoding/json/marshal.odin:297]: Remove the indentation and {} for each case block.

**Comment by Skytrias** [comment]
nice thank you for the review, forgot about the switch styling :+1:

**Review by gingerBill — APPROVED** [review]

**Pull request merged**